core/store/src/trie/split_state.rs

```diff
@@ -0,0 +1,373 @@
+use crate::trie::iterator::TrieItem;
+use crate::{get_delayed_receipt_indices, set, ShardTries, StoreUpdate, Trie, TrieUpdate};
+use borsh::BorshDeserialize;
+use bytesize::ByteSize;
+use near_primitives::account::id::AccountId;
+use near_primitives::errors::StorageError;
+use near_primitives::receipt::Receipt;
+use near_primitives::shard_layout::ShardUId;
+use near_primitives::trie_key::TrieKey;
+use near_primitives::types::{StateChangeCause, StateRoot};
+use std::collections::HashMap;
+use std::str::from_utf8;
+
+impl Trie {
+    /// Computes the set of trie items (nodes with keys and values) for a state part.
+    ///
+    /// # Panics
+    /// storage must be a TrieCachingStorage
+    /// part_id must be in [0..num_parts)
+    ///
+    /// # Errors
+    /// StorageError if the storage is corrupted
+    pub fn get_trie_items_for_part(
+        &self,
+        part_id: u64,
+        num_parts: u64,
+        state_root: &StateRoot,
+    ) -> Result<Vec<TrieItem>, StorageError> {
+        assert!(part_id < num_parts);
+        assert!(self.storage.as_caching_storage().is_some());
+
+        let path_begin = self.find_path_for_part_boundary(state_root, part_id, num_parts)?;
+        let path_end = self.find_path_for_part_boundary(state_root, part_id + 1, num_parts)?;
+        self.iter(state_root)?.get_trie_items(&path_begin, &path_end)
+    }
+}
+
+impl ShardTries {
+    /// Apply `changes` to build states for new shards
+    /// `state_roots` contains state roots for the new shards
+    /// Ignore changes on DelayedReceipts or DelayedReceiptsIndices
+    /// Update `store_update` and return new state_roots
+    /// used for building states for new shards in resharding
+    pub fn apply_changes_to_new_states<'a>(
+        &self,
+        state_roots: HashMap<ShardUId, StateRoot>,
+        changes: Vec<(Vec<u8>, Option<Vec<u8>>)>,
+        key_to_shard_id: &(dyn Fn(&[u8]) -> Result<Option<ShardUId>, StorageError> + 'a),
+    ) -> Result<(StoreUpdate, HashMap<ShardUId, StateRoot>), StorageError> {
+        let mut changes_by_shard: HashMap<_, Vec<_>> = HashMap::new();
+        for (raw_key, value) in changes.into_iter() {
+            if let Some(new_shard_uid) = key_to_shard_id(&raw_key)? {
+                assert!(
+                    state_roots.contains_key(&new_shard_uid),
+                    "Cannot find state roots for shard {:?} for key {}. Only has state roots for shards {:?}",
+                    new_shard_uid,
+                    from_utf8(&raw_key).unwrap(),
```

Comment:
Good point. I am going to move this assert outside of this function for a clearer error message. This error will happen if `ShardLayout` is not specified correctly. Since we specify which shards will be split to which shards in shard layout, if that information is not consistent with the result of `account_id_to_shard_id`, this assert will fail.